src/PROPOSAL/detail/PROPOSAL/Secondaries.cxx

```diff
@@ -225,23 +236,30 @@ std::shared_ptr<ParticleState> Secondaries::GetExitPoint(
     return nullptr; // No exit point found
 }
 
-std::shared_ptr<ParticleState> Secondaries::GetClosestApproachPoint(const Geometry& geometry) const
+std::shared_ptr<ParticleState> Secondaries::GetClosestApproachPoint(
+        const Geometry& geometry) const
 {
-    for (unsigned int i = 0; i < track_.size(); i++) {
-        auto sec_pos = track_.at(i).position;
-        auto sec_dir = track_.at(i).direction;
-        if (geometry.DistanceToClosestApproach(sec_pos, sec_dir) <= 0.) {
-            if(std::abs(geometry.DistanceToClosestApproach(sec_pos, sec_dir))
-                        < PARTICLE_POSITION_RESOLUTION)
+   if (track_.size() == 1)
+       return std::make_unique<ParticleState>(track_.front());
+
+    for (unsigned int i = 0; i < track_.size() - 1; i++) {
+        auto pos_i = track_.at(i).position;
+        auto pos_f = track_.at(i+1).position;
+
+        auto displacement = pos_f - pos_i;
+        auto dist_i_f = displacement.magnitude();
+        displacement.normalize();
+
+        auto distance_to_closest_approach
+            = geometry.DistanceToClosestApproach(pos_i, displacement);
+        if (distance_to_closest_approach == dist_i_f) {
+            return std::make_unique<ParticleState>(track_.at(i+1));
+        } else if (distance_to_closest_approach < dist_i_f) {
+            if (distance_to_closest_approach < PARTICLE_POSITION_RESOLUTION)
                 return std::make_unique<ParticleState>(track_.at(i));
-            if (i == 0)
-                return std::make_unique<ParticleState>(track_.front());
-            auto prev_pos = track_.at(i-1).position;
-            auto prev_dir = track_.at(i-1).direction;
-            auto displacement = geometry.DistanceToClosestApproach(prev_pos,
-                                                                   prev_dir);
-            auto closest_approach = RePropagateDistance(track_.at(i-1),
-                                                        displacement);
+
+            auto closest_approach = RePropagateDistance(
+                    track_.at(i), displacement, distance_to_closest_approach);
```

Comment:
Are these bounds checks, which are also always computed in `GetEntryPoint` and `GetExitPoint`, really necessary? Maybe using `[]` instead of `.at()` would increase the performance.

Reply:
Since it is safe that the index exist inside the loop (as long at the header is correct), we should be fine to use `[]`